pkg/actor/resize_pvc.go

```diff
@@ -0,0 +1,144 @@
+		return errors.Wrap(err, "failed to fetch statefulset")
+	}
+
+	if statefulSetIsUpdating(statefulSet) {
+		return NotReadyErr{Err: errors.New("statefulset is updating, waiting for the update to finish")}
+	}
+
+	if len(statefulSet.Spec.VolumeClaimTemplates) == 0 {
+		log.Info("Skipping PVC resize as PVCs do not exist")
+		return nil
+	}
+
+	stsStorageSizeDeployed := statefulSet.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()
+	stsStorageSizeSet := cluster.Spec().DataStore.VolumeClaim.PersistentVolumeClaimSpec.Resources.Requests.Storage()
+
+	if stsStorageSizeDeployed.Equal(stsStorageSizeSet.DeepCopy()) {
+		log.Info("Skipping PVC resize as sizes match")
+		return nil
+	}
+
+	log.Info("Starting PVC resize")
+
+	return rp.findAndResizePVC(ctx, statefulSet, cluster)
+}
+
+// findAndResizePVC finds all active PVCs and resizes them to the current size
+func (rp *resizePVC) findAndResizePVC(ctx context.Context, sts *appsv1.StatefulSet, cluster *resource.Cluster) error {
+	// K8s doesn't provide a way to tell if a PVC or PV is currently in use by
+	// a pod. However, it is safe to assume that any PVCs with an ordinal great
+	// than or equal to the sts' Replicas is not in use. As only pods with with
+	// an ordinal < Replicas will exist. Any PVCs with an ordinal less than
+	// Replicas is in use. To detect this, we build a map of PVCs that we
+	// consider to be in use and skip and PVCs that it contains
+	// the name of.
+	prefixes := make([]string, len(sts.Spec.VolumeClaimTemplates))
+	pvcsToKeep := make(map[string]bool, int(*sts.Spec.Replicas)*len(sts.Spec.VolumeClaimTemplates))
+	for j, pvct := range sts.Spec.VolumeClaimTemplates {
+		prefixes[j] = fmt.Sprintf("%s-%s-", pvct.Name, sts.Name)
+
+		for i := int32(0); i < *sts.Spec.Replicas; i++ {
+			name := fmt.Sprintf("%s-%s-%d", pvct.Name, sts.Name, i)
+			pvcsToKeep[name] = true
+		}
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
+	if err != nil {
+		return errors.Wrap(err, "converting statefulset selector to metav1 selector")
+	}
+
+	clientset, err := kubernetes.NewForConfig(rp.config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create kubernetes clientset")
+	}
+
+	pvcs, err := clientset.CoreV1().PersistentVolumeClaims(cluster.Namespace()).List(ctx, metav1.ListOptions{
+		LabelSelector: selector.String(),
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "finding PVCs to for resizing")
+	}
+
+	for _, pvc := range pvcs.Items {
+		// Resize PVCs that are still in use
+		if pvcsToKeep[pvc.Name] {
+			pvc.Spec.Resources.Requests.Storage().Set(cluster.Spec().DataStore.VolumeClaim.PersistentVolumeClaimSpec.Resources.Requests.Storage().MilliValue())
+			if _, err := clientset.CoreV1().PersistentVolumeClaims(cluster.Namespace()).Update(ctx, &pvc, metav1.UpdateOptions{}); err != nil {
+				return errors.Wrap(err, "error resizing PVCs")
+			}
+		}
+	}
+
+	log := rp.log.WithValues("CrdbCluster", cluster.ObjectKey())
+	log.Info("resized all PVCs")
```

Comment:
Don't we also need to restart all the CRDB pods to get the PVCs to resize as well as modifying the statefulset's template to include the increased size?

Reply:
Yes we do.  Forgot to make this WIP.